### ISSUE TITLE
mediaListEndReached Workaround?

### ISSUE THREAD

**Issue by McSmoove** [caprica/vlcj#540]
Greetings.

I am trying to implement a videoPlaylist playback having more playlists to be played.

I wanted to start playing a new playlist when the previous one ended but as far as i tested looking examples and documentations seems like the method mediaListEndReached ( witch should be raised when all the media was played by the player, i guess ) never fires.

I wanted to ask if there is some kiind of workaround to be 100% sure that i reached the end of my media playlist in order to start a new one.

Thank you

**Comment by caprica** [comment]
If the event never fires, that's probably a bug.

Are you playing the playlist in loop mode?

In vlcj, all of the native event handling is pretty much the same - so if the event is not firing I would suspect it's not firing from the native code rather than vlcj code. Although it's possible there's a vlcj bug in relation to this, at the moment I doubt it.

As for workarounds, count the number of entries in your playlist, and count how many have played. Not ideal of course.

Personally, I always try and maintain playlists in Java rather than relying on the native media list functionality.

**Comment by McSmoove** [comment]
My original idea is to play the playlist in loop if i have no new playlists and have a global boolean that is true if i ended my playlist only one time even if it was looped X times, and if i have a new playlist comming and the previous ended at least one time i update the mediaList with the new contents.

I was thinking for the workaround that you suggested but i didnt liked it, but i guess that is the only way :|

**Comment by caprica** [comment]
I'm just thinking if your are playing in loop mode, the end will never be reached - hence no event.

**Comment by McSmoove** [comment]
I will count every finished event of the current media, and if it is greater than the length of my playlist i suppose that i reached the end of the playlist at least one time.

It sucks that you monitor the single media that is ending but can't monitor the ending of an mediaList.

So the workaround is to divide the mediaList in single media, count the times that you play the single media and than suppose that is it is greater than the mediaList size you reached at least one time the end of your mediaList

**Comment by caprica** [comment]
Before you commit to a workaround, have you actually conclusively shown that the end event is not fired, at least if you don't play in loop mode. Because it would be simple to loop the media list yourself - when you get the end event you play the list again.

If the event never fires under those circumstances, there's surely a bug in LibVLC/VLC or maybe (but much less likely imho) vlcj.

**Comment by McSmoove** [comment]
Yes, if you addMediaListEventListener and Override mediaListEndReached from MediaListEventListener, that event is never raised when you reach the end of playing your mediaList without Looping

At the same time the events mediaListWillDeleteItem, mediaListWillAddItem, mediaListItemDeleted, mediaListItemAdded works very well

**Comment by caprica** [comment]
OK, I will do a check some time soon-ish to see if there's a vlcj bug. I'm sceptical though. If it's a LibVLC/VLC bug then that is a much harder thing to track down.

**Comment by McSmoove** [comment]
I am not saying that is a vlcj bug, im just saying that this thing don't work so other people can know :)

Thank you anyway :)

**Comment by caprica** [comment]
So I appreciate this issue is almost 2 years old now, but there is an update...

A long time ago in a galaxy far, far away, longer than even when this issue was created, earlier versions of LibVLC had no way to tell when a media list was finished - i.e .when the last item in the list had finished playing. There was also an unused libvlc_MediaListPlayerPlayed event. I missed this change in VLC, a commit some few years ago repurposed that unused event so that it fires for exactly the case in point.

i.e. libvlc_MediaListPlayerPlayed will now fire when the last item in a media list has finished playing.

The name of the event is kinda awkward and confusing, but nevertheless this is the event we want. I always thought that even would be fired at the beginning, i.e. "when the media list player was played", but nope.

I only discovered this as part of my major refactoring for vlcj-4.x.

So, in short, this event will be exposed as some sort of semantic "playlist finished playing the last item" event in vlcj-4.x.

**Comment by caprica** [comment]
Some other notes for the record...

This event will NOT fire after the last item has been played if the media list player is in LOOP mode (as you'd expect).

Nor will this even fire if the media list player is in REPEAT mode (again, as you'd expect).

There are some idiosyncrasies...

In REPEAT mode, if you invoke play on the media list player, you will receive this played event immediately - without any media playing. It seems when in REPEAT mode you can only use the API that plays a specific item index (which kinda makes sense I suppose).

**Comment by caprica** [comment]
This has been address as part of the major refactoring of media list players committed under dbaeed3c7df32ae4a6f086824462770ee2e4fa25.
